utils/dyn-config.c

```diff
@@ -0,0 +1,597 @@
+
+  option = glusterBlockGetOption(p);
+  if (!option) {
+    r = s;
+    while (isblank(*r)) {
+      r++;
+    }
+    LOG("dyncfg", GB_LOG_INFO, "option type s:'%s', r:'%s' not supported!\n", s, r);
+
+    if (isdigit(*r)) {
+      type = GB_OPT_INT;
+    } else {
+      type = GB_OPT_STR;
+    }
+
+    option = glusterBlockRegisterOption(p, type);
+    if (!option) {
+      return;
+    }
+  }
+
+  /* parse the int/string type options */
+  switch (option->type) {
+  case GB_OPT_INT:
+    while (!isdigit(*s)) {
+      s++;
+    }
+    r = s;
+    while (isdigit(*r)) {
+      r++;
+    }
+    *r= '\0';
+
+    option->optInt = atoi(s);
+    break;
+  case GB_OPT_STR:
+    while (isblank(*s)) {
+      s++;
+    }
+    /* skip first " or ' if exist */
+    if (*s == '"' || *s == '\'') {
+      s++;
+    }
+    r = q - 1;
+    while (isblank(*r)) {
+      r--;
+    }
+    /* skip last " or ' if exist */
+    if (*r == '"' || *r == '\'') {
+      *r = '\0';
+    }
+
+    /* free if this is reconfig */
+    if (option->optStr) {
+      GB_FREE(option->optStr);
+    }
+    GB_STRDUP(option->optStr, s);
+    break;
+  default:
+    LOG("dyncfg", GB_LOG_ERROR,
+        "option type %d not supported!\n", option->type);
+    break;
+  }
+}
+
+static void
+glusterBlockParseOptions(gbConfig *cfg, char *buf, int len, bool reloading)
+{
+  char *cur = buf, *end = buf + len;
+
+  while (cur < end) {
+    /* skip blanks lines */
+    GB_SKIP_BLANK_LINES(cur, end);
+
+    /* skip comments with '#' */
+    GB_SKIP_COMMENT_LINES(cur, end);
+
+    if (cur >= end) {
+      break;
+    }
+
+    if (!isalpha(*cur)) {
+      continue;
+    }
+
+    /* parse the options from config file to gb_options[] */
+    glusterBlockParseOption(&cur, end);
+  }
+
+  /* parse the options from gb_options[] to struct gbConfig */
+  glusterBlockConfSetOptions(cfg, reloading);
+}
+
+static int
+glusterBlockLoadConfig(gbConfig *cfg, bool reloading)
+{
+  int ret = -1;
+  int fd, len;
+  char *buf;
+
+  buf = CALLOC(GB_MAX_CFG_FILE_SIZE);
+  if (!buf) {
+    return -ENOMEM;
+  }
+
+  fd = open(cfg->configPath, O_RDONLY);
+  if (fd < 0) {
+    LOG("dyncfg", GB_LOG_ERROR,
+        "Failed to open file '%s', %m\n", cfg->configPath);
+    goto free_buf;
+  }
+
+  len = glusterBlockReadConfig(fd, buf, GB_MAX_CFG_FILE_SIZE);
+  close(fd);
+  if (len < 0) {
+    LOG("dyncfg", GB_LOG_ERROR,
+        "Failed to read file '%s'\n", cfg->configPath);
+    goto free_buf;
+  }
+
+  buf[len] = '\0';
+
+  glusterBlockParseOptions(cfg, buf, len, reloading);
+
+  ret = 0;
+free_buf:
+  GB_FREE(buf);
+  return ret;
+}
+
+#define BUF_LEN 1024
+static void *
+glusterBlockDynConfigStart(void *arg)
+{
+  gbConfig *cfg = arg;
+  int monitor, wd, len;
+  char buf[BUF_LEN];
+
+  monitor = inotify_init();
+  if (monitor == -1) {
+    LOG("dyncfg", GB_LOG_ERROR,
+        "Failed to init inotify %d\n", monitor);
+    return NULL;
+  }
+
+  wd = inotify_add_watch(monitor, cfg->configPath, IN_ALL_EVENTS);
```

Comment:
I think masking IN_MODIFY should be good enough, why do we need watch for IN_ALL_EVENTS ?

Reply:
Yeah, there was one story about this when testing this in tcmu-runner's dynamic reloading, long time ago, but currently I couldn't recall it. Just could remember it won't work in some corner case.